sdk/storage/Azure.Storage.Blobs/src/Models/GetBlobs.cs

```diff
@@ -0,0 +1,288 @@
+        /// Check if two GetBlobsOptions instances are equal.
+        /// </summary>
+        /// <param name="obj">The instance to compare to.</param>
+        /// <returns>True if they're equal, false otherwise.</returns>
+        public bool Equals(GetBlobsOptions other) =>
+            this.IncludeCopyMetadata == other.IncludeCopyMetadata &&
+            this.IncludeDeletedBlobs == other.IncludeDeletedBlobs &&
+            this.IncludeMetadata == other.IncludeMetadata &&
+            this.IncludeSnapshots == other.IncludeSnapshots &&
+            this.IncludeUncommittedBlobs == other.IncludeUncommittedBlobs &&
+            this.Prefix == other.Prefix &&
+            this.PageSizeHint == other.PageSizeHint;
+    }
+
+    internal class GetBlobsAsyncCollection : StorageAsyncCollection<BlobItem>
+    {
+        private readonly BlobContainerClient _client;
+        private readonly GetBlobsOptions? _options;
+
+        public GetBlobsAsyncCollection(
+            BlobContainerClient client,
+            GetBlobsOptions? options,
+            CancellationToken cancellationToken)
+            : base(cancellationToken)
+        {
+            this._client = client;
+            this._options = options;
+            this.PageSizeHint = options?.PageSizeHint;
+        }
+
+        protected override async Task<Page<BlobItem>> GetNextPageAsync(
+            string continuationToken,
+            bool isAsync,
+            CancellationToken cancellationToken = default)
+        {
+            var task = this._client.GetBlobsAsync(
+                continuationToken,
+                this._options,
+                this.PageSizeHint,
+                isAsync,
+                cancellationToken);
+            var response = isAsync ?
+                await task.ConfigureAwait(false) :
+                task.EnsureCompleted(syncOverAsync: true);
+            return new Page<BlobItem>(
+                response.Value.BlobItems.ToArray(),
+                response.Value.NextMarker,
+                response.GetRawResponse());
+        }
+    }
+
+    internal class GetBlobsByHierarchyAsyncCollection : StorageAsyncCollection<BlobHierarchyItem>
+    {
+        private readonly BlobContainerClient _client;
+        private readonly GetBlobsOptions? _options;
+        private readonly string _delimiter;
+
+        public GetBlobsByHierarchyAsyncCollection(
+            BlobContainerClient client,
+            string delimiter,
+            GetBlobsOptions? options,
+            CancellationToken cancellationToken)
+            : base(cancellationToken)
+        {
+            this._client = client;
+            this._delimiter = delimiter;
+            this._options = options;
+            this.PageSizeHint = options?.PageSizeHint;
+        }
+
+        protected override async Task<Page<BlobHierarchyItem>> GetNextPageAsync(
+            string continuationToken,
+            bool isAsync,
+            CancellationToken cancellationToken = default)
```

Comment:
We should not use default cancellationToken in internal APIs.

Reply:
Yes - fixed.  I was toying with the idea of making this part of the default interface for AsyncCollection but decided it's probably not easy to generalize across services.